### PR TITLE
Base code for Lambda Function

### PR DESCRIPTION
**Pull request by anchen1011**
This section of code is the base function for Lambda Inference. LIT as a tool will modify `model_service.py` and `model_url.yaml` for generalized purpose.

### CODE REVIEW

File: lit-base/model_service.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+def postprocess(output):
+    """
+    Postprocess model output into response body.
+    :return: response body string
+    output : NDArray
+        output from mx mod
+    """
+    # load labels from synset
+    with open('synset.txt', 'r') as f:
```
[the remark this answers is not in the document]
Comment (anchen1011) [reply lit-base/model_service.py]: deleted for duplication

File: lit-base/model_service.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    """
+    img = None
+    try:
+        # GET method
+        if event.has_key('httpMethod') and event['httpMethod'] == 'GET':
+            url = event['queryStringParameters']['url']
+            # download image from url
+            img_file = tempfile.NamedTemporaryFile(delete=True)
+            if url:
+                req = urllib2.urlopen(url)
```
[the remark this answers is not in the document]
Comment (anchen1011) [reply lit-base/model_service.py]: Here we are trying to use tmp file to avoid garbage handling. We may move to mx.test_utils once we decide how we are going to post-clean exactly.

File: lit-base/model_loader.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    for k, v in save_dict.items():
+        tp, name = k.split(':', 1)
+        if tp == 'arg':
+            arg_params[name] = v
+        if tp == 'aux':
+            aux_params[name] = v
+    return symbol, arg_params, aux_params
+
+
+model_url = yaml.load(open("model_url.yaml", "r"))
```
[the remark this answers is not in the document]
Comment (anchen1011) [reply lit-base/model_loader.py]: The reason why we intentionally avoid config.py is that our toolkit may modify the config automatically, which is well-supported for .yaml but not .py

File: lit-base/model_service.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    """
+    Preprocess Lambda event into input format for mx mod.
+    :return: input NDArray
+    event : dict
+        Lambda event
+    """
+    img = None
+    try:
+        # GET method
+        if event.has_key('httpMethod') and event['httpMethod'] == 'GET':
```
[the remark this answers is not in the document]
Comment (anchen1011) [reply lit-base/model_service.py]: Modified accordingly

File: lit-base/lambda_function.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
+from model_service import preprocess, inference, postprocess
+
+
+def lambda_handler(event, context):
+
+    # serving 
+    input    = preprocess(event)
```
[the remark this answers is not in the document]
Comment (anchen1011) [reply lit-base/lambda_function.py]: Modified accordingly

File: lit-base/lambda_function.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+    output   = inference(input)
+    response = postprocess(output)
+    
+    out = {
```
[the remark this answers is not in the document]
Comment (anchen1011) [reply lit-base/lambda_function.py]: Modified accordingly

File: lit-base/model_loader.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    arg_params = {}
+    aux_params = {}
+    for k, v in save_dict.items():
+        tp, name = k.split(':', 1)
+        if tp == 'arg':
+            arg_params[name] = v
+        if tp == 'aux':
+            aux_params[name] = v
+    return symbol, arg_params, aux_params
+
```
[the remark this answers is not in the document]
Comment (anchen1011) [reply lit-base/model_loader.py]: Modified accordingly

File: lit-base/model_loader.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+        Model parameter, dict of name to NDArray of net's weights.
+    aux_params : dict of str to NDArray
+        Model parameter, dict of name to NDArray of net's auxiliary states.
+    """
+    symbol = mx.symbol.load(s_fname)
+    save_dict = mx.nd.load(p_fname)
+    arg_params = {}
+    aux_params = {}
+    for k, v in save_dict.items():
+        tp, name = k.split(':', 1)
```
[the remark this answers is not in the document]
Comment (anchen1011) [reply lit-base/model_loader.py]: Modified accordingly

File: lit-base/model_service.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (anchen1011) [reply lit-base/model_service.py]: Modified accordingly

File: lit-base/model_service.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+def postprocess(output):
+    """
+    Postprocess model output into response body.
+    :return: response body string
+    output : NDArray
+        output from mx mod
+    """
+    # load labels from synset
+    with open('synset.txt', 'r') as f:
+        labels = [l.rstrip() for l in f]
```
[the remark this answers is not in the document]
Comment (anchen1011) [reply lit-base/model_service.py]: Modified accordingly

File: lit-base/model_service.py [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+
+def inference(input):
+    """
+    Run inference on input.
+    :return: output NDArray
+    input : NDArray
+        input for mx mod
+    """
+    # define a simple data batch
+    from collections import namedtuple
```
[the remark this answers is not in the document]
Comment (anchen1011) [reply lit-base/model_service.py]: Modified accordingly

File: lit-base/model_service.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+    Batch = namedtuple('Batch', ['data'])
```
[the remark this answers is not in the document]
Comment (anchen1011) [reply lit-base/model_service.py]: Slightly special in namedtuple case - https://docs.python.org/2/library/collections.html#collections.namedtuple

File: lit-base/model_service.py [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+    # load required models
+    mod = mx.mod.Module(symbol=sym, context=ctx, label_names=None)
+    mod.bind(for_training=False, data_shapes=[('data', input.shape)], 
+             label_shapes=mod._label_shapes)
+    mod.set_params(arg_params, aux_params, allow_missing=True)
+    # run inference and get the output
+    mod.forward(Batch([input]))
+    prob = mod.get_outputs()[0].asnumpy()
+    prob = np.squeeze(prob)
+    a = np.argsort(prob)[::-1]
```
[the remark this answers is not in the document]
Comment (anchen1011) [reply lit-base/model_service.py]: Modified accordingly

File: lit-base/model_url.yaml [hunk 13]
```diff
@@ -0,0 +1,2 @@
+url_params: "https://lambda-inference-toolkit-test.s3.amazonaws.com/resnext-101-64x4d-0000.params"
```
[the remark this answers is not in the document]
Comment (anchen1011) [reply lit-base/model_url.yaml]: (I deleted some duplicated comments)

File: lit-base/lambda_function.py [hunk 14]
```diff
@@ -0,0 +1,24 @@
+from model_service import preprocess, inference, postprocess
+
+
+class BaseResponse:
```
Comment (kalyc) [inline lit-base/lambda_function.py]: 1. move to a another file
2. Make it inherit a dict - 

class BaseResponse(dict): 
 ....

this way a dict will be returned

File: lit-base/lambda_function.py [hunk 15]
```diff
@@ -0,0 +1,24 @@
+from model_service import preprocess, inference, postprocess
```
Comment (kalyc) [inline lit-base/lambda_function.py]: rename file to lambda_handler

File: lit-base/lambda_function.py [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+
+
+class BaseResponse:
+
+    def __init__(self, response_body):
+        self.headers = {"content-type": "application/json", "Access-Control-Allow-Origin": "*"}
+        self.statusCode = 200
+        self.body = response_body
+
+    def get_dict(self):
```
Comment (kalyc) [inline lit-base/lambda_function.py]: see comment above. this can be removed

File: lit-base/lambda_function.py [hunk 17]
```diff
@@ -0,0 +1,24 @@
+from model_service import preprocess, inference, postprocess
+
+
+class BaseResponse:
+
+    def __init__(self, response_body):
+        self.headers = {"content-type": "application/json", "Access-Control-Allow-Origin": "*"}
```
Comment (kalyc) [inline lit-base/lambda_function.py]: add dict.__init__(self) - once you inherit from dict

File: lit-base/lambda_function.py [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+        self.statusCode = 200
+        self.body = response_body
+
+    def get_dict(self):
+        return {"headers":self.headers, "body":self.body, "statusCode":self.statusCode}
+
+
+def lambda_handler(event, context):
+
```
Comment (kalyc) [inline lit-base/lambda_function.py]: [remove line]

File: lit-base/lambda_function.py [hunk 19]
```diff
@@ -0,0 +1,24 @@
...
+    # serving 
+    data    = preprocess(event)
+    output   = inference(data)
+    response_body = postprocess(output)
+    
+    response = BaseResponse(response_body).get_dict()
```
Comment (kalyc) [inline lit-base/lambda_function.py]: response = BaseResponse(postprocess(output))  - after making above modifications

File: lit-base/model_service.py [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+             label_shapes=mod._label_shapes)
+    mod.set_params(arg_params, aux_params, allow_missing=True)
+    # run inference and get the output
+    mod.forward(Batch([data]))
+    prob = mod.get_outputs()[0].asnumpy()
+    prob = np.squeeze(prob)
+    return np.argsort(prob)[::-1][0]
+
+
+def postprocess(output):
```
Comment (kalyc) [inline lit-base/model_service.py]: can postprocess() be a private method? & be invoked within inference() itself?

File: lit-base/model_loader.py [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+    aux_params = {}
+    for key, value in save_dict.items():
+        tp, name = key.split(':', 1)
+        if tp == 'arg':
+            arg_params[name] = value
+        if tp == 'aux':
+            aux_params[name] = value
+    return symbol, arg_params, aux_params
+
+model_url = yaml.load(open("model_url.yaml", "r"))
```
Comment (kalyc) [inline lit-base/model_loader.py]: please rename `model_url` file to `config`

File: lit-base/model_url.yaml (hunk 13 above)
[the remark this answers is not in the document]
Comment (kalyc) [reply lit-base/model_url.yaml]: consider renaming to 'params' & 'symbol' instead of 'url_params' & 'url_symbol'

File: lit-base/lambda_function.py (hunk 15 above)
Comment (anchen1011) [reply lit-base/lambda_function.py]: lambda_function.py is Lambda standard

File: lit-base/lambda_function.py (hunk 14 above)
Comment (anchen1011) [reply lit-base/lambda_function.py]: Different from Java, the correct way to organize Python file is catering readability instead of enforcing one class per file. 

Actually after considering carefully, I'm not sure about the idea of creating a class now. Let's wait a bit for more inputs.

File: lit-base/lambda_function.py (hunk 18 above)
Comment (anchen1011) [reply lit-base/lambda_function.py]: Modified accordingly

Comment (anchen1011) [reply lit-base/lambda_function.py]: modified accordingly

File: lit-base/model_loader.py [hunk 22]
```diff
@@ -0,0 +1,50 @@
+import yaml
+
+import mxnet as mx
+
+import tempfile
+from urllib import urlretrieve
+
+
+# load model
+def load_model(s_fname, p_fname):
```
[the remark this answers is not in the document]
Comment (anchen1011) [reply lit-base/model_loader.py]: Modified accordingly

File: lit-base/model_loader.py (hunk 21 above)
Comment (anchen1011) [reply lit-base/model_loader.py]: The `yaml` is exactly for `model_url`. I'm very not sure about this renaming.

File: lit-base/model_loader.py [hunk 23]
```diff
@@ -0,0 +1,50 @@
...
+f_params_file_name = f_params_file.name
+
+# download symbol
+f_symbol_file = tempfile.NamedTemporaryFile(delete=True)
+urlretrieve(url_symbol, f_symbol_file.name)
+f_symbol_file.flush()
+f_symbol_file_name = f_symbol_file.name
+
+# load model from symbol and params
+sym, arg_params, aux_params = load_model(f_symbol_file_name, f_params_file.name)
```
[the remark this answers is not in the document]
Comment (anchen1011) [reply lit-base/model_loader.py]: Modified accordingly

File: lit-base/model_service.py (hunk 20 above)
Comment (anchen1011) [reply lit-base/model_service.py]: We are considering make inference a handler. But usually standard abstraction is preprocess, postprocess and inference. In any way postprocess should not be treated differently than preprocess.

File: lit-base/model_url.yaml (hunk 13 above)
[the remark this answers is not in the document]
Comment (anchen1011) [reply lit-base/model_url.yaml]: The concern is that they are exactly URLs.

File: lit-base/model_service.py [hunk 24]
```diff
@@ -0,0 +1,92 @@
...
+                req = urllib2.urlopen(url)
+                img_file.write(req.read())
+                img_file.flush()
+            img = Image.open(img_file.name)
+            img_file.close()
+        # POST method
+        else:
+            multipart_string = event['body'].decode('base64')
+            img = Image.open(StringIO(multipart_string))
+    except KeyError:
```
[the remark this answers is not in the document]
Comment (anchen1011) [reply lit-base/model_service.py]: Modified accordingly

File: lit-base/model_service.py [hunk 25]
```diff
@@ -0,0 +1,91 @@
...
```
[the remark this answers is not in the document]
Comment (anchen1011) [reply lit-base/model_service.py]: Modified accordingly

File: lit-base/lambda_function.py (hunk 15 above)
Comment (kalyc) [reply lit-base/lambda_function.py]: I don't see any documentation on setting that as the standard name. Would recommend giving it a meaningful name - `model_service_invoke.py` perhaps?

Comment (anchen1011) [reply lit-base/lambda_function.py]: Thanks for the suggestion! I insist `lambda_function` is more meaningful here for two reason:
1. It's the default option for lambda function in both Get Started and Template
2. It's more important for implementor to find the entry point of lambda function, which is defined in Lambda configuration not the lambda package, than to give the file a usage specific name

File: lit-base/lambda_function.py (hunk 14 above)
Comment (anchen1011) [reply lit-base/lambda_function.py]: I went over this with MAR folks, and to support MAR we will prefer a BaseResponse. However, inherit from dict should be reconsidered carefully for two reasons:
1. HTTP headers is not a map, it's a list of key value pair. We need a way to format it into dictionary so that Lambda can consume. 
2. http://neethack.com/2017/04/Why-inheritance-is-bad/

File: lit-base/lambda_function.py [hunk 26]
```diff
@@ -0,0 +1,24 @@
...
+    def __init__(self, response_body):
+        self.headers = {"content-type": "application/json", "Access-Control-Allow-Origin": "*"}
+        self.statusCode = 200
+        self.body = response_body
+
+    def get_dict(self):
+        return {"headers":self.headers, "body":self.body, "statusCode":self.statusCode}
+
+
+def lambda_handler(event, context):
```
Comment (nswamy) [inline lit-base/lambda_function.py]: Please add comments to **every method/class** on what it is doing ?
follow the format similar to 
https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/executor.py#L113

File: lit-base/model_loader.py (hunk 22 above)
Comment (nswamy) [inline lit-base/model_loader.py]: follow the documentation standard similar to 
https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/executor.py#L113

File: lit-base/model_loader.py (hunk 21 above)
Comment (nswamy) [reply lit-base/model_loader.py]: these are configurable parameters that is passed to the function. `model_config` makes sense
I think the point that @kalyc is making is easy you can easily extend this if you want to more configuration parameters like `batch_size`, etc,. 
Also why would you need this yaml file if you use model archive

File: lit-base/model_loader.py [hunk 27]
```diff
@@ -0,0 +1,50 @@
...
+            arg_params[name] = value
+        if tp == 'aux':
+            aux_params[name] = value
+    return symbol, arg_params, aux_params
+
+model_url = yaml.load(open("model_url.yaml", "r"))
+url_params = model_url["url_params"]
+url_symbol = model_url["url_symbol"]
+
+# download params
```
Comment (nswamy) [inline lit-base/model_loader.py]: can you make a single method `download_file(url)` that returns the file name and just call for both symbol and params

File: lit-base/model_service.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nswamy) [reply lit-base/model_service.py]: @anchen1011 While I understand NamedTempFile removes the boiler plate code to automatically handles deletion of file upon close. I suggest you consider using `mx. test_utils.download` code to remove a dependency.

File: lit-base/model_service.py [hunk 28]
```diff
@@ -0,0 +1,91 @@
...
+        Lambda event
+    """
+    img = None
+    try:
+        # GET method
+        if event.get('httpMethod') == 'GET':
+            url = event['queryStringParameters']['url']
+            # download image from url
+            img_file = tempfile.NamedTemporaryFile(delete=True)
+            if url:
```
Comment (nswamy) [inline lit-base/model_service.py]: its fine for a demo(not asking to remove right now) 

should you support url? 
might create an opportunity for someone to just pass a URL and DDOS this service. All I need is a small lambda function that creates 100s of requests point at some large binary. You can bring down two services the one which hosts the binary and this lambda.
@kalyc @frankfliu what do you think?

File: lit-base/model_service.py [hunk 29]
```diff
@@ -0,0 +1,91 @@
...
+import urllib2
+import numpy as np
+from PIL import Image
+
+from model_loader import sym, arg_params, aux_params
+
+from collections import namedtuple
+from StringIO import StringIO
+
+
```
Comment (nswamy) [inline lit-base/model_service.py]: Do you plan to retain this code when you support MAR? If not lets skip reviewing this.

File: lit-base/model_url.yaml (hunk 13 above)
[the remark this answers is not in the document]
Comment (nswamy) [reply lit-base/model_url.yaml]: IMO spelling that the config is in fact a `url` is a good thing and makes it easy to read/understand code.

File: lit-base/lambda_function.py (hunk 14 above)
Comment (nswamy) [reply lit-base/lambda_function.py]: @kalyc `BaseResponse` is not `dict`,  You should use Inheritance when your classes can fit `is a` paradigm ie., The idea behind it is `old code using new code`. If a user has written code that uses a base class, he can just pass the new improved Derived class without much code change.

File: lit-base/model_service.py (hunk 28 above)
Comment (nswamy) [reply lit-base/model_service.py]: you should leave it to the client code to fetch from the host/service and it pass it you.

File: README.md [hunk 30]
```diff
@@ -0,0 +1,64 @@
+# Lambda-Inference-Toolkit (LIT)
```
[the remark this answers is not in the document]
Comment (anchen1011) [reply README.md]: Thanks. I will reorganize all documentations

File: lit-base/lambda_function.py (hunk 26 above)
Comment (anchen1011) [reply lit-base/lambda_function.py]: Sure thanks! Also, for BaseResponse->HTTPResponse should we try to avoid using the same name as https://docs.python.org/2.2/lib/httpresponse-objects.html or it is fine?

File: lit-base/model_service.py [hunk 31]
```diff
@@ -0,0 +1,91 @@
...
+    """
+    img = None
+    try:
+        # GET method
+        if event.get('httpMethod') == 'GET':
+            url = event['queryStringParameters']['url']
+            # download image from url
+            img_file = tempfile.NamedTemporaryFile(delete=True)
+            if url:
+                req = urllib2.urlopen(url)
```
Comment (apeforest) [inline lit-base/model_service.py]: I think we need to wrap a retry logic here to make it more robust

File: lit-base/model_service.py [hunk 32]
```diff
@@ -0,0 +1,91 @@
...
+        else:
+            multipart_string = event['body'].decode('base64')
+            img = Image.open(StringIO(multipart_string))
+    except KeyError:
+        # direct invocation
+        url = event['url']
+        # download image from url
+        img_file = tempfile.NamedTemporaryFile(delete=True)
+        if url:
+            req = urllib2.urlopen(url)
```
Comment (apeforest) [inline lit-base/model_service.py]: add retry here. Even better, create a retry block to reuse.

File: lit-base/model_loader.py (hunk 21 above)
Comment (anchen1011) [reply lit-base/model_loader.py]: @nswamy @apeforest Modified accordingly

File: lit-base/model_loader.py (hunk 27 above)
Comment (anchen1011) [reply lit-base/model_loader.py]: Modified accordingly in lit-mar

File: lit-base/model_service.py (hunk 28 above)
Comment (anchen1011) [reply lit-base/model_service.py]: @nswamy Agree! Modified accordingly in lit-mar

File: lit-base/model_service.py [hunk 33]
```diff
@@ -0,0 +1,91 @@
...
+    event : dict
+        Lambda event
+    """
+    img = None
+    try:
+        # GET method
+        if event.get('httpMethod') == 'GET':
+            url = event['queryStringParameters']['url']
+            # download image from url
+            img_file = tempfile.NamedTemporaryFile(delete=True)
```
[the remark this answers is not in the document]
Comment (anchen1011) [reply lit-base/model_service.py]: Modified accordingly

File: lit-base/model_service.py [hunk 34]
```diff
@@ -0,0 +1,91 @@
...
+            img_file.close()
+        # POST method
+        else:
+            multipart_string = event['body'].decode('base64')
+            img = Image.open(StringIO(multipart_string))
+    except KeyError:
+        # direct invocation
+        url = event['url']
+        # download image from url
+        img_file = tempfile.NamedTemporaryFile(delete=True)
```
[the remark this answers is not in the document]
Comment (anchen1011) [reply lit-base/model_service.py]: Modified accordingly

File: lit-base/model_service.py (hunk 31 above)
Comment (anchen1011) [reply lit-base/model_service.py]: I slightly prefer fail the request directly under Lambda context for two reasons:
1. Lambda has limited time and resources to execute, and should succeed or fail fast when it's triggered
2. Retry could easily expose customer to extra charge than they expect, and make them vulnerable to malicious attack

File: lit-base/lambda_function.py (hunk 14 above)
Comment (kalyc) [reply lit-base/lambda_function.py]: @nswamy `BaseResponse` is structured as a dictionary - there are key value pairs associated with the status code, headers, body. With the current implementation, a method explicitly needs to be added `get_dict` and this route can be avoided. Hence the suggestion above.

File: lit-base/model_service.py (hunk 31 above)
Comment (apeforest) [reply lit-base/model_service.py]: Adding a retry block does not violate the Lambda resource limitation. It will make this service more robust. E.g. in the case of overloading, you apparently just don't want to just break downstream service right away.

Comment (anchen1011) [reply lit-base/model_service.py]: Thanks! Modified accordingly. Since concern 

2. Retry could easily expose customer to extra charge than they expect, and make them vulnerable to malicious attack

is still valid, I only sleep it for a short time and retry once

File: lit-base/model_service.py [hunk 35]
```diff
@@ -28,7 +28,11 @@ def preprocess(event):
             # download image from url
             with tempfile.NamedTemporaryFile(delete=True) as img_file:
                 if url:
-                    req = urllib2.urlopen(url)
+                    try:
+                        req = urllib2.urlopen(url)
+                    except urllib2.URLError:
+                        time.sleep(0.01)
+                        req = urllib2.urlopen(url)
```
Comment (apeforest) [inline lit-base/model_service.py]: Thanks for adding the retry logic here. I think instead of looping using the ```except``` clause it might be more intuitive to use a while loop. That way you can also make the number of retries more flexible.

File: lit-base/model_service.py [hunk 36]
```diff
@@ -42,7 +46,11 @@ def preprocess(event):
         # download image from url
         with tempfile.NamedTemporaryFile(delete=True) as img_file:
             if url:
-                req = urllib2.urlopen(url)
+                try:
+                    req = urllib2.urlopen(url)
+                except urllib2.URLError:
+                    time.sleep(0.01)
+                    req = urllib2.urlopen(url)
```
Comment (apeforest) [inline lit-base/model_service.py]: Same for here too

File: lit-base/model_service.py (hunk 35 above)
Comment (nswamy) [reply lit-base/model_service.py]: Also can you make the retry counter and retry interval configurable with defaults. So the User can control whether he wants a retry, how long or just set the counter to 1 and you don't retry.

For the case of Lambda, I think retry adds additional latency and the user wouldn't know if it is because a Model file is throttled by S3 or model unavailable intermittently due to the hosting, but we can let the user decide using configuration.

Comment (nswamy) [reply lit-base/model_service.py]: pseudo-code for exponential backoff. 
```
retry = 0
while (retry++ < retrycount):
	get_url
if Not available:
	sleep( baseMilliseconds * 2^retry)
```

**Comment by apeforest** [comment]
LGTM after discussion about the customized retry option. Good job!

File: base/model_loader.py [hunk 37]
```diff
@@ -0,0 +1,87 @@
...
+    for key, value in save_dict.items():
+        tp, name = key.split(':', 1)
+        if tp == 'arg':
+            arg_params[name] = value
+        if tp == 'aux':
+            aux_params[name] = value
+    return symbol, arg_params, aux_params
+
+
+def download_url(url, target, retries=2, base_retry_interval=0.01):
```
Comment (nswamy) [inline base/model_loader.py]: you need to take retries and base_retry_interval from a config file. Please do it as a separate PR.
